webapp/packages/core-authentication/src/AuthInfoService.ts

```diff
@@ -29,57 +30,58 @@ export class AuthInfoService {
     private readonly userInfoResource: UserInfoResource,
     private readonly authProvidersResource: AuthProvidersResource,
     private readonly windowsService: WindowsService,
+    private readonly asyncTaskInfoService: AsyncTaskInfoService,
   ) {}
 
-  login(providerId: string, options: ILoginOptions): ITask<UserInfo | null> {
-    return new AutoRunningTask(async () => await this.userInfoResource.login(providerId, options)).then(authInfo =>
-      this.federatedAuthentication(providerId, options, authInfo),
-    );
+  async login(providerId: string, options: ILoginOptions): Promise<UserInfo | null> {
+    await this.userInfoResource.login(providerId, options);
+    return this.userInfoResource.data;
   }
 
-  private federatedAuthentication(
-    providerId: string,
-    options: ILoginOptions,
-    { redirectLink, authId, authStatus }: AuthInfo,
-  ): ITask<UserInfo | null> {
-    let window: Window | null = null;
-    let id = providerId;
+  asyncLogin(providerId: string, options: IAsyncLoginOptions): ITask<UserInfo | null> {
+    const task = this.asyncTaskInfoService.create(async () => {
+      const result = await this.userInfoResource.asyncAuthLogin(providerId, options);
+
+      let window: Window | null = null;
+      let id = providerId;
 
-    if (options.configurationId) {
-      const configuration = this.authProvidersResource.getConfiguration(providerId, options.configurationId);
+      if (options.configurationId) {
+        const configuration = this.authProvidersResource.getConfiguration(providerId, options.configurationId);
 
-      if (configuration) {
-        id = configuration.id;
+        if (configuration) {
+          id = configuration.id;
+        }
       }
-    }
 
-    if (redirectLink) {
-      id = uuid();
-      window = this.windowsService.open(id, {
-        url: redirectLink,
-        target: id,
-        width: 600,
-        height: 700,
-      });
+      if (result.redirectLink) {
+        id = uuid();
+        window = this.windowsService.open(id, {
+          url: result.redirectLink,
+          target: id,
+          width: 600,
+          height: 700,
+        });
 
-      if (window) {
-        window.focus();
+        if (window) {
+          window.focus();
+        }
       }
-    }
+
+      return result.taskInfo;
+    });
 
     return new AutoRunningTask(
-      () => {
-        if (authId && authStatus === AuthStatus.InProgress) {
-          return this.userInfoResource.finishFederatedAuthentication(authId, options.linkUser);
-        }
+      async () => {
+        const info = await this.asyncTaskInfoService.run(task);
+        await this.userInfoResource.getAuthTaskResult(info.id);
 
-        return AutoRunningTask.resolve(this.userInfoResource.data);
-      },
-      () => {
         if (window) {
           this.windowsService.close(window);
         }
```

Comment:
This is the global window object, i think something completely wrong here

Reply:
Variable shadowing, fixed